### PR TITLE
Live Preview of Dynamic Data

### LINKED ISSUE #1181: Displaying live preview of Dynamic Values
### Description:

Some way to display a live preview of the data in Dynamic Values.

Keeping in mind that we can't mess with the DOM to update the value of the text itself.

### PR DESCRIPTION
**Pull request by HardeepAsrani**
Closes #1181.

### Summary

This PR should allow us to show a live preview of dynamic values in the editor. It's, currently, made for Single Posts so let me know how it works for you and how you'd like to see it extend to support Custom Layouts better.

This also adds Post Content field to Dynamic Conditions.

### Screenshots 
https://www.loom.com/share/6b9fa68030fd4539b6c07176501a2c61

---- 

### Checklist before the final review

- [x] Visual elements are not affected by independent changes.
- [x] It is at least compatible with the [minimum WordPress version](https://wordpress.org/plugins/otter-blocks/).
- [x] It loads additional script in frontend only if it is required.
- [x] Does not impact the [Core Web Vitals](https://web.dev/vitals/).
- [x] In case of deprecation, old blocks are safely migrated.
- [x] It is usable in Widgets and FSE.

### CODE REVIEW

**Comment by HardeepAsrani** [comment]
@JohnPixle @mghenciu Feel free to review it.

**Comment by irinelenache** [comment]
@HardeepAsrani Did some testing and found a single thing (i think it's supposed to work like this until the functionality is extended): 

- [ ] The dynamic preview doesn't work for a query loop https://vertis.d.pr/WocWA7

Besides this, everything seems to work fine with the preview 🚀

**Comment by HardeepAsrani** [comment]
@irinelenache Yes, kind of, We don't have a way to know the ID of the block in loop.

**Comment by mghenciu** [comment]
> @JohnPixle @mghenciu Feel free to review it.

Looks good Hardeep, and I like the switching mechanics between picking the Dynamic Value and the Preview.
**I am thinking it's better to differentiate those 2 in terms of UI,** so my suggestion would be to leave the Dynamic Values as it is (discrete Otter Palette in background), and when in view it's rendered a Dynamic Content (show that content with **2 Underlines**, using the Otter main color) - similarly how Kadence does this with the [Dynamic values](https://www.kadencewp.com/blog/how-to-use-dynamic-content-with-kadence/): 
image


___
And one thing I noticed is after adding a lot of Dynamic Tags, there's some kind of flick when trying to select values, so I have to click twice (I am on Chrome, macOS):

https://user-images.githubusercontent.com/52494172/192744250-cbf9a139-6107-4c25-b42c-0aedebaa93ce.mov



https://we-soldier-seya.instawp.xyz/wp-admin
weviwohuzo9685
uUMHDQFWkqbosltic3hp

**Comment by Soare-Robert-Daniel** [comment]
@HardeepAsrani, I do not see some performance bottlenecks.

One micro performance will be to replace the third-party hash function with the following function:

```javascript
const hashCode = ( str ) => str.split( '' ).reduce( ( s, c ) => Math.imul( 31, s ) + c.charCodeAt( 0 ) | 0, 0 ).toString();
```

The performance:
image

Test code:
```javascript
const s1 = performance.now();
const attrsHash = hash( attrs );
const e1 = performance.now();

const s2 = performance.now();
const newHash = hashCode( JSON.stringify( attrs ) );
const e2 = performance.now();

console.group( 'Object' );
console.log( `Obj: ${JSON.stringify( attrs )}` );

console.group( 'Library object-hash' );
console.log( `Hash: ${attrsHash}` );
console.log( `Time: ${e1 - s1}` );
console.groupEnd();

console.group( 'Function' );
console.log( `Hash: ${newHash}` );
console.log( `Time: ${e2 - s2}` );
console.groupEnd();

console.groupEnd();

File: src/blocks/plugins/dynamic-content/value/index.js [hunk 1]
```diff
@@ -38,3 +51,68 @@ export const format = {
...
+						displayWaitlist[ attrsHash ] = [ element ];
+
+						const interval = setInterval( () => {
+							value = select( 'themeisle-gutenberg/data' ).getDynamicData( attrs );
+
+							if ( undefined !== value ) {
+								clearInterval( interval );
+								displayWaitlist[ attrsHash ].forEach( el => displayData( el, value ) );
+							}
+						}, 5000 );
```
Comment (Soare-Robert-Daniel) [inline src/blocks/plugins/dynamic-content/value/index.js:100]: I think we can put a limit on how many times this interval can run for a hash code.

Comment (Soare-Robert-Daniel) [reply src/blocks/plugins/dynamic-content/value/index.js:100]: Also, you can delete the items after loading.

```javascript
displayWaitlist[ attrsHash ] = [];
```

No need to keep them in memory.

**Comment by HardeepAsrani** [comment]
@Soare-Robert-Daniel Your function works fine normally but in certain cases with just a minor variation, like object keys being in a different order, it gives a different hash.

**Comment by HardeepAsrani** [comment]
@mghenciu The flickering happens because of the data change. Once you click on the dynamic text first, it just transforms from the preview to the actual text so you have to click again to put it on focus. Can't do anything about it.

Also I've added the underline but kept it to `var(--wp-admin-theme-color)` instead as Otter colors made it look like it was a grammatical error with the red/orangish line. 😄

**Pull request merged**